File: Days.30/Python/1.Linked-List.py

class Node:
    def __init__(self,val=-1):
        self.val=val
        self.next=None
class LinkedList:

    def __init__(self):
        self.head=None
        self.tail=None

    def InsertEnd(self,val):
        if self.head==None:
            self.head=self.tail=Node(val)
        else:
            self.tail.next=Node(val)
            self.tail=self.tail.next

    def InsertBegin(self,val):
        if self.head==None:
            self.head= self.tail =Node(val)
        else:
            temp=Node(val)
            temp.next=self.head
            self.head=temp

    def InsertPosition(self,val,pos):
        #Pos ----> Based on 1-Based Indexing
        n=self.size()
        if pos<0 or pos>n+1:
            print('Index Out of Bound')
            return
        else:
            if pos==1:
                self.InsertBegin(val)
            elif pos==n+1:
                self.InsertEnd(val)
            else:
                prev=None
                curr=self.head
                i=1
                while i!=pos:
                    prev=curr
                    curr=curr.next
                    i+=1
                temp=Node(val)
                temp.next=curr
                prev.next=temp

        
    def DeletionEnd(self):
        if self.head==None:
            return -1
        if self.head.next==None:
            val=self.head.val
            self.head=self.tail=None
            return val
        else:
            temp=self.head
            prev=None
            while temp.next!=None:
                prev=temp
                temp=temp.next
            val=temp.val
            prev.next=None
            return val
        
    def DeletionBegin(self):
        if self.head==None:
            return -1
        if self.head.next==None:
            val=self.head.val
            self.head=tail=None
            return val
        val=self.head.val
        self.head=self.head.next 
        return val
        
    def DeletionPosition(self,pos):
        if self.head==None:
            return -1
        if pos<0 or pos>self.size():
            return -1
        if pos==1:
            return self.DeletionBegin()
        elif pos==self.size():
            return self.DeletionEnd()
        else:
            i=1
            prev=None
            curr=self.head
            while i!=pos:
                prev=curr
                curr=curr.next
                i+=1
            val=curr.val
            prev.next=curr.next 
            return val

    def size(self):
        sz=0
        temp=self.head
        while temp!=None:
            sz+=1
            temp=temp.next
        return sz

def main():
    linkedlist=LinkedList()
    while True:
        print('1.Insert at the End')
        print('2.Insert at the Beginning')
        print('3.Insert at Given Position')
        print('4.Deletion at the End')
        print('5.Deletion at the Beginning')
        print('6.Deletion at Given Position')
        print('7.Size of Linked List')
        print('8.Print')
        print('9.Exit')
        choice=int(input('Enter Your Choice: '))
        if choice==1:
            val=int(input('Enter a Value: '))
            linkedlist.InsertEnd(val)
        
        elif choice==2:
            val=int(input('Enter a Value: '))
            linkedlist.InsertBegin(val)
        
        elif choice==3:
            val=int(input('Enter a Value: '))
            pos=int(input('Enter a Position: '))
            linkedlist.InsertPosition(val,pos)
        
        elif choice==4:
            val=linkedlist.DeletionEnd()
            if val==-1:
                print('Linked List is Empty')
            else:
                print('Deleted Value:',val)
        
        elif choice==5:
            val=linkedlist.DeletionBegin()
            if val==-1:
                print('Linked List is Empty')
            else:
                print('Deleted Value:',val)
        
        elif choice==6:
            pos=int(input('Enter a Position: '))
            val=linkedlist.DeletionPosition(pos)
            if val==-1:
                print('Index Out of Bound')
            else:
                print('Deleted Value:',val)
        
        elif choice==7:
            print('Size:',linkedlist.size())

        elif choice==8:
            if linkedlist.head==None:
                print('Linked List is Empty')
            else:
                temp=linkedlist.head
                while temp!=None:
                    print(temp.val,end=" ")
                    temp=temp.next
                print()
        elif choice==9:
            break
        else:
            print('Invalid Choice')
    

main()

File: Days.30/C++/1.Linked-List.cpp
#include<iostream>
#include<vector>

using namespace std;


class Node
{
    public:
    int val;
    Node *next;
    Node(int val)
    {
        this->val=val;
        this->next=NULL;
    }
};

class LinkedList
{
    public:
    Node *head;
    Node *tail;
    LinkedList()
    {
        head=NULL;
        tail=NULL;
    }
    int size()
    {
        int sz=0;
        Node *temp=head;
        while(temp!=NULL)
        {
            sz+=1;
            temp=temp->next;
        }
        return sz;
    }
    void InsertBegin(int val)
    {
        if(head==NULL)
        {
            head=tail=new Node(val);
        }
        else
        {
            Node *temp=new Node(val);
            temp->next=head;
            head=temp;
        }
    }

    void InsertEnd(int val)
    {
        if(head==NULL)
        {
            head=tail=new Node(val);
        }
        else
        {
            tail->next=new Node(val);
            tail=tail->next;
        }
    }

    void InsertPosition(int val,int pos)
    {
        int n=size();
        if(pos<0 or pos>n+1)
        {
            cout<<"Index Out of Bound\n";
        }
        else
        {
            if(pos==1)
            {
                InsertBegin(val);
            }
            else if (pos==n+1)
            {
                InsertEnd(val);
            }
            else
            {
                Node *prev,*curr;
                prev=NULL;
                curr=head;
                int i=1;
                while(i!=pos)
                {
                    prev=curr;
                    curr=curr->next;
                    i+=1;
                }
                Node *temp=new Node(val);
                temp->next=curr;
                prev->next=temp;
            }
        }

    }

    int DeleteBegin()
    {
        if(head==NULL)
        {
            return -1;
        }
        else{
            int val=head->val;
            Node *temp=head->next;
            delete head;
            head=temp;
            return val;
        }
    }

    int DeleteEnd()
    {
        if(head==NULL)
        {
            return -1;
        }
        else if(head->next==NULL)
        {
            int val=head->val;
            Node *temp=head;
            head=head->next;
            delete temp;
            return val;
        }
        else{
            Node *curr,*prev;
            prev=NULL;
            curr=head;
            while(curr->next!=NULL)
            {
                prev=curr;
                curr=curr->next;
            }
            int val=curr->val;
            prev->next=NULL;
            delete curr;
            return val;
        }
    }

    int DeletePosition(int pos)
    {
        int n=size();
        if (pos<0 or pos>n)
        {
            return -1;
        }
        if(pos==1)
        {
            return DeleteBegin();
        }
        else if(pos==n)
        {
            return DeleteEnd();
        }
        else{
            Node *curr,*prev;
            curr=head;
            prev=NULL;
            int i=1;
            while(i!=pos)
            {
                prev=curr;
                curr=curr->next;
            }
            int val;
            val=curr->val;
            prev->next=curr->next;
            delete curr;
            return val;
        }
    }

    void print()
    {
        Node *temp=head;
        while(temp!=NULL)
        {
            cout<<temp->val<<" ";
            temp=temp->next;
        }
        cout<<"\n";
    }
};


int main()
{
    LinkedList *linkedlist=new LinkedList();
    int choice,val,pos;
    while(true)
    {
        cout<<"1.Insert at the Beginning\n";
        cout<<"2.Insert at the End\n";
        cout<<"3.Insert at Given Position\n";
        cout<<"4.Delete at the Beginning\n";
        cout<<"5.Delete at the End\n";
        cout<<"6.Delete at Given Position\n";
        cout<<"7.Size\n";
        cout<<"8.Print List\n";
        cout<<"9.Exit\n";
        cout<<"Enter Your Choice:\t";
        cin>>choice;
        if(choice==1)
        {
            cout<<"Enter a Value:\t";
            cin>>val;
            linkedlist->InsertBegin(val);
        }
        else if(choice==2)
        {
            cout<<"Enter a Value:\t";
            cin>>val;
            linkedlist->InsertEnd(val);
        }
        else if(choice==3)
        {
            cout<<"Enter a Value:\t";
            cin>>val;
            cout<<"Enter a Position:\t";
            cin>>pos;
            linkedlist->InsertPosition(val,pos);
        }
        else if(choice==4)
        {
            val=linkedlist->DeleteBegin();
            if (val==-1)
            {
                cout<<"List is Empty\n";
            }
            else{
                cout<<"Deleted Value:\t"<<val<<"\n";
            }
        }
        else if(choice==5)
        {
            val=linkedlist->DeleteEnd();
            if(val==-1)
            {
                cout<<"List is Empty\n";
            }
            else{
                cout<<"Deleted Value:\t"<<val<<"\n";
            }
        }
        else if(choice==6)
        {
            cout<<"Enter a Position:\t";
            cin>>pos;
            val=linkedlist->DeletePosition(pos);
            if(val==-1)
            {
                cout<<"Index Out of Bound\n";
            }
            else
            {
                cout<<"Deleted Value at "<<pos<<": "<<val<<"\n";
            }
        }
        else if(choice==7)
        {
            cout<<"Size:\t"<<linkedlist->size()<<"\n";
        }
        else if(choice==8)
        {
            linkedlist->print();
        }
        else if(choice==9)
        {
            break;
        }
        else
        {
            cout<<"Invalid Choice\n";
        }
    }
    
    return 0;
}

